Add EOS design document (#191)
E3SM-Project/Omega: components/omega/doc/design/EOS.md

(omega-design-eos)=

# Equation of State

> ## 1 Overview The equation of state relates density to the prognostic state

Reviewer: start sentence on new line after the heading for proper formatting

variables temperature and salinity.  The prognostic temperature, $\theta$, is
either potential temperature or Conservative Temperature, depending on the
chosen equation of state, and  $S$ is the salinity.  Given  that the governing
equations are  non-Boussinesq, the equation of state class or functor will
provide specific volume from the state variables.  It will provide methods for
computing the specific volume, and its first derivatives.

## 2 Requirements

### 2.1 Requirement: calculate specific volume/density
The specific volume and density will be calculated form from prognostic
temperature, salinity, and pressure (ref function `gsw_specvol(sa,ct,p`)).

### 2.2 Requirement: allow a choice of equations of state
Omega-1 should include the option for two EOS options. The first is a linear
eos, for benchmarking against MPAS-O and to use in simplified cases. The other
option should be a well-established version of TEOS-10 (e.g. Roquet 75-term
polynomial). 

### 2.3 Requirement: computational efficiency

> The eos code should not be major computational bottleneck (criteria?)

Reviewer: Bottleneck is unrealistic. In every ocean model I've worked with, the EOS is one of the 2-3 top kernels and is always a bottleneck. TEOS-10 is even more expensive with a larger polynomial. Maybe something more like: As one of the most expensive kernels, the EOS must be as efficient as possible. Developers should be discouraged from calling the EOS too frequently.

Reviewer 2: Agreed. As you can tell, I was trying to convey *an* expectation of reasonable performance, but struggled to find the right wording/criteria. I appreciate your suggestion.

### 2.4 Requirement: abide by license for the GSW Oceanographic Toolbox
The GSW toolbox can only be used without modification, so a Omega-specific port
of certain functions will be required to support Kokkos/GPU enabled
computation.  The toolbox without modifications can still be used for testing.

### 2.5 Requirement: verification 
The Omega implementation should be compared with the test value published in
Roquet et al 2015, and the official GSW Oceanographic Toolbox (supported by
TEOS-10) to verify correctness.

### 2.6 Desired: freezing temperature calculation
Later versions should include calculating the freezing temperature of seawater.

> ### 2.7 Desired: output in situ temperature

Reviewer: Is this really a function for the EOS?

Later versions should include calculating the in situ  temperature needed for
coupling (at surface including at non-zero pressure).

### 2.8 Desired: output first derivatives
The EOS will need to be able to produce alpha and beta (drho/dT and drho/dS)

> for linear expansions in the higher-order pressure gradient.

Reviewer: and for some mixing parameterizations.

### 2.9 Desired: check value range
Later versions should include a check on the range of ocean properties to

> assess use of TEOS-10 (equivalent to the ``ocean funnel'' check).

Reviewer: This is really a hard requirement I think? If the input ocean model quantities exceed the valid range for the polynomial, the deviations can be quite extreme. And how do we want to treat out-of-range values that can unfortunately arise in a typical ocean simulation? In previous models, we have typically truncated the values to the valid range only for the purposes of the EOS calculation while leaving the values unchanged otherwise.

Reviewer: btw, this requirement also makes the point-wise evaluation a little problematic. In prior implementations, it was often more efficient to do the range restriction as part of some pre-processing together with some other preparatory calculations before hitting the polynomial loop.

> ### 2.10 Desired: check value range

Reviewer: duplicate heading - replace with input conversion?

Later versions should include functions to convert between conservative and
potential temperatures, and absolute and practical salinity. These will be used
to convert the initial conditions or compare to ocean states from other sources
(MPAS-Ocean, reanalysis, other models etc.) These functions may be used offline
in pre- or post-processing but need to be consistent with the EOS
implementation. 

> 

Reviewer: I believe we still have a requirement for adiabatically displaced density for the purposes of computing stability or Brunt-Vaisala freq. For stability, this is often from neighboring vertical layers. For mixed-layer models, we sometimes need surface-displaced densities.

## 3 Algorithmic Formulation

## 4 Design
The GSW-C toolbox will be incorporated into Omega as a submodule.
In order to abide by the license of the toolbox, the toolbox will only be used
in testing to check our implementation. The coefficients for the Roquet et al.
2015 expansion will be based on the published values (Appendix). A Kokkos
implementation of the function to compute specific volume will be ported to
Omega. The GSW-C toolbox sub-module will serve as a baseline reference for our
ports in unit tests. 

> For flexibility in optimizing performance, calls to the EOS function should be

Reviewer: As mentioned above, I suspect we want to restrict computing density to one time per step and given the EOS complexity, probably won't be inlined or merged with other kernels. Guess we shouldn't design it out, but the preference would be whole-array if there are any compromises associated with the single-cell support.

device, EOS functions will be provided as functors with a consistent interface.
Functions that use the EOS (e.g. pressure gradient) will have a template
argument for the EOS type. A runtime switch statement is then used to call the
user function with the EOS selected in the configuration file.  This approach
allows the EOS to be called within parallel loops without any inner logic
statements and maintains the flexibility to calculate the EOS for a single cell
so EOS calculations can be fused with other computation.  If performance can be
improved by pre-calculating the EOS, the functors can be used to provide
whole-array calculations functions as well.

### 4.1 Data types and parameters
An `enum class` will be used to specify options for the EOS used for an Omega
simulation:
```c++
enum class EOSType{
   Linear,  /// Linear equation of state
   TEOS10Poly75t.  /// Roquet et al. 2015 75 term expansion
}
```
The user will select the EOS type at runtime in the input yaml file under an
eos section:
```yaml
    eos:
       eosType: 'teos10'
```
### 4.2 Functions
Both the linear and TEOS-10 EOS options will have a functor that provides the
same interface:
```c++

> class LinearEOS {

Reviewer: Do we really need a separate class for each EOS choice? I think we should have just a single EOS class (see related comment later)

  Real operator()(...);
};
class TEOS10Poly75t {
  Real operator()(...); // same arguments as LinearEOS
};
```

> Any function that depends on the EOS will be a template:

Reviewer: Similarly here - not sure we need a templated class based on EOS type.

```c++
template <class EOSType>
void userOfEOS(EOSType EOS, ...) {
  parallelFor(...) {
    Real Rho = EOS(...);
  }
}
```

> Functions that depend on the EOS will be called as follows:

Reviewer: EOS choice should be encapsulated in the class and not exposed external to the EOS. I would prefer a single EOS interface with this choice/conditional embedded inside rather than exposed to external calling routines.

As one of the most expensive kernels, the EOS must be as efficient as possible. Developers should be discouraged from calling the EOS too frequently. The fraction of compute time used be the EOS within Omega1 will be compared to fraction used in MPAS-Ocean. 

### 2.4 Requirement: abide by license for the GSW Oceanographic Toolbox
The GSW toolbox can only be used without modification, so a Omega-specific port of certain functions will be required to support Kokkos/GPU enabled computation. The toolbox without modifications can still be used for verification testing against the Omega implementation.

### 2.5 Requirement: verification
The Omega implementation should be compared with the test value published in Roquet et al 2015, and the official GSW Oceanographic Toolbox (supported by TEOS-10) to verify correctness.

### 2.6 Requirement: check value range
The EOS code should include a check on the range of ocean properties to assess use of TEOS-10 (equivalent to the ``ocean funnel'' check), and a truncation of state variables to use the polynomial only in the valid range (similar to the truncation in MPAS-Ocean). 

### 2.7 Requirement: provide adiabatically-displaced density

> The EOS code should be able to be used to calculate density adiabatically displaced to i) the layer above, ii) the surface. This is needed to calculate the Brunt-Vaisala and for mixed-layer parameterizations. This operation should be done as efficiently as possible (e.g. can we reuse parts of the polynomial? this may also be a candidate for Kernel Fusion?).

Reviewer 3: maybe just me and I'm fine to not change, but having questions in the design doc is odd.  maybe say (e.g., exploration of reusing parts of the polynomial or kernel fusion)

Reviewer 2: True, this was more of an internal comment than a necessary description. Thanks for the suggestion.

> The EOS code should be able to be used to calculate density adiabatically displaced to i) the layer above, ii) the surface. This is needed to calculate the Brunt-Vaisala and for mixed-layer parameterizations. This operation should be done as efficiently as possible (e.g. can we reuse parts of the polynomial? this may also be a candidate for Kernel Fusion?).

Reviewer 2: ```suggestion
The EOS code should be able to be used to calculate density adiabatically displaced to i) the layer below, ii) the surface. This is needed to calculate the Brunt-Vaisala and for mixed-layer parameterizations. This operation should be done as efficiently as possible (e.g. we will explore reusing parts of the polynomial or leveraging Kernel Fusion).
```

### 2.8 Desired: freezing temperature calculation
Later versions should include calculating the freezing temperature of seawater.

### 2.9 Desired: output in situ temperature
Later versions should include calculating the in situ  temperature needed for coupling (at surface including at non-zero pressure). 

### 2.10 Desired: output first derivatives
The EOS will need to be able to produce alpha and beta (drho/dT and drho/dS) for linear expansions in the higher-order pressure gradient and for some mixing parameterizations.

### 2.11 Desired: variable conversion
Later versions should include functions to convert between conservative and potential temperatures, and absolute and practical salinity. These will be used to convert the initial conditions or compare to ocean states from other sources (MPAS-Ocean, reanalysis, other models etc.) These functions may be used offline in pre- or post-processing but need to be consistent with the EOS implementation. 

### 2.12 Desired: enthalpy conversion
Later versions should include functions to calculate the layer potential enthalpy from enthalpy fluxes to convert coupling fluxes to changes in conservative temperature in a rigorous manner. 

> ## 3 Algorithmic Formulation

Reviewer 3: can we add a quick statement about algorithmic formulations aren't needed as they are described by ... (TEOS-10 doc)?

> ## 3 Algorithmic Formulation

Reviewer 2: ```suggestion
## 3 Algorithmic Formulation
The formulation of the specific volume calculation will be the 75-term polynomial as documented in Roquet et al. 2015. The implementation of the displaced density calculation may be altered to improve performance by reusing some polynomial coefficients, but should produce the same results as the full 75-term polynomial. 
```

```c++
void Eos::erase(const std::string &Name);
void Eos::clear();
```
## 5 Verification and Testing

### 5.1 Test: Verification of GSW-C submodule
A unit test will verify that the result of the GSW-C toolbox (used as a submodule, without modifications) matches the expected value of specific volume published in Roquet et al. 2015 within machine precision. The publication only includes one data point (Sa, Ct, P) to check. 

### 5.2 Test: Verification of our TEOS-10 75-term polynomial implementation
A unit test will verify that the result of the Omega implementation of the Roquet et al. 2015 75-term polynomial calculation of the specific volume matches the expected value of specific volume published in Roquet et al. 2015 within machine precision. The publication only includes one data point (Sa, Ct, P) to check. 

### 5.3 Test: Verification of TEOS-10 with GSW-C toolbox
A unit test will verify that the result of the Omega and GSW-C toolbox implementations for the Roquet et al. 2015 expansion are within machine precision over a range of conservative temperature, absolute salinity, and pressure ranges.

### 5.4 Test: Verification of linear EOS

> The linear EOS should be close, but not equal (to machine precision)  to the TEOS-10 result. It should be exactly linear in T,S. It will be verified against known values from the MPAS-Ocean model for a range of temperature and salinity values.

Reviewer 4: @sbrus89 and @alicebarthel, several questions come to mind here:
* When you say, "The linear EOS should be close, but not equal (to machine precision)  to the TEOS-10 result", do you mean for a reference T and S around which the EOS has been linearized?
* When you say "it should be exactly linear in T,S", is "it" the specific volume?
* If yes, a linear specific-volume EOS is not the same as a linear density EOS, so how will the comparison with MPAS-Ocean (which is linear in density) be performed?

Reviewer 2: Thanks Xylar!

-  Yes, we typically do a check for the calculated value to match the published value (to 10^-10 a the moment) for the reference T, S input. A quick check for the linear EOS was that for the same input values, the calculated specific volume should be "close to" (e.g. 10^-2) the published value, but not of equal precision as TEOS-10 (10^-10). That was for a single value so maybe not that valuable but it was a mental check I did (i.e. a linear EOS sanity check) so I just added it in the unit tests.
- You are right, the "it" is density. (not specific volume)
- As mentioned above, the linear EOS implemented is an eos that is linear in density. It returns specific volume but the linearity (and check for linearity) is on the density. Which can be compared to MPAS-Ocean.

> The linear EOS should be close, but not equal (to machine precision)  to the TEOS-10 result. It should be exactly linear in T,S. It will be verified against known values from the MPAS-Ocean model for a range of temperature and salinity values.

Reviewer 2: ```suggestion
The linear EOS should be close, but not equal to the TEOS-10 result. The calculated density should be exactly linear in T,S. It will be verified against known values from the MPAS-Ocean model for a range of temperature and salinity values.
```
